src/System.Text.Encodings.Web.Utf8/UrlEncoder.cs

```diff
@@ -1,19 +1,336 @@
+                currentDecodeBits = byte1.Value & 0x07;
+                byteCount = 4;
+                expectValueMin = 0x10000;
+            }
+            else
+            {
+                // invalid first byte
+                return false;
+            }
+
+            var remainingBytes = byteCount - 1;
+            while (remainingBytes > 0)
+            {
+                // read following three chars
+                if (reader == buffer.Length)
+                {
+                    return false;
+                }
+
+                var nextItr = reader;
+                var nextByte = UnescapePercentEncoding(ref nextItr, end, buffer);
+                if (!nextByte.HasValue)
+                {
+                    return false;
+                }
+
+                if ((nextByte & 0xC0) != 0x80)
+                {
+                    // the follow up byte is not in form of 10xx xxxx
+                    return false;
+                }
+
+                currentDecodeBits = (currentDecodeBits << 6) | (nextByte.Value & 0x3F);
+                remainingBytes--;
+
+                if (remainingBytes == 1 && currentDecodeBits >= 0x360 && currentDecodeBits <= 0x37F)
+                {
+                    // this is going to end up in the range of 0xD800-0xDFFF UTF-16 surrogates that
+                    // are not allowed in UTF-8;
+                    return false;
+                }
+
+                if (remainingBytes == 2 && currentDecodeBits >= 0x110)
+                {
+                    // this is going to be out of the upper Unicode bound 0x10FFFF.
+                    return false;
+                }
+
+                reader = nextItr;
+                if (byteCount - remainingBytes == 2)
+                {
+                    byte2 = nextByte.Value;
+                }
+                else if (byteCount - remainingBytes == 3)
+                {
+                    byte3 = nextByte.Value;
+                }
+                else if (byteCount - remainingBytes == 4)
+                {
+                    byte4 = nextByte.Value;
+                }
+            }
+
+            if (currentDecodeBits < expectValueMin)
+            {
+                // overlong encoding (e.g. using 2 bytes to encode something that only needed 1).
+                return false;
+            }
+
+            // all bytes are verified, write to the output
+            if (byteCount > 0)
+            {
+                buffer[writer++] = (byte)byte1;
+            }
+            if (byteCount > 1)
+            {
+                buffer[writer++] = (byte)byte2;
+            }
+            if (byteCount > 2)
+            {
+                buffer[writer++] = (byte)byte3;
+            }
+            if (byteCount > 3)
+            {
+                buffer[writer++] = (byte)byte4;
+            }
+
+            return true;
+        }
+
+        private static void Copy(int begin, int end, ref int writer, Span<byte> buffer)
```

Comment:
The CoreCLR version does, but the corefxlab one doesn't https://github.com/dotnet/coreclr/blob/SpanOfT/src/mscorlib/src/System/Span.cs#L352.